[LLR-Phase1]: Implementation of the LLR  metod into Grid

fbonnet08 opened the pull request:
This is the first phase of the LLR implementatino into Grid-Telos. Coda has been tested and several tests are included. `Test_llr_contrained_action.cc`, the launcher for this test is found in `./scripts/llr_tools/Run_Test_llr_contrained_action_nodes001_mpi01-01-01-04.sh` the call to the readcommandline will be removed once the pull request has been accepted.

fbonnet08 committed:
- [Update]: Removal of ``` In file included from /pfs/lustrep1/users/bonnetfr/SwanSea/SourceCodes/Grid-Telos/Grid/Grid/threads/Pragmas.h:32: /pfs/lustrep1/users/bonnetfr/SwanSea/SourceCodes/Grid-Telos/Grid/Grid/threads/Accelerator.h:481:14: warning: 'hipMallocHost' is deprecated: use hipHostMalloc instead [-Wdeprecated-declarations]   auto err = hipMallocHost((void **)&ptr,bytes);              ^ /opt/rocm-6.0.3/include/hip/hip_runtime_api.h:2956:1: note: 'hipMallocHost' has been explicitly marked deprecated here DEPRECATED("use hipHostMalloc instead") ^ ``` Warning message.

fbonnet08 committed:
- [Dev]: insertion of the instantiation in GenericHMCrunner.h.
- Testcode for the HMC.

Commit:
-- .ac/C++ ---
- [Dev]: Insertion of front end --enable-LLR in configure.ac.
- [Dev]: Instantiation of the GenericHMCRunner in HMC build directory.
--------------

Commit:
-- .ac/C++ ---
- [Dev]: code update in configure.ac.
- [Dev]: code edit.
--------------

Commit:
-- .ac/C++ ---
- [Dev]: lower casing of the front end switches.
- Testing of the compilation on the new switches and verification of the instantiation in the main code LLR_HMC_WilsonGauge.cc code.
--------------

Commit:
-- C++ -------
- [Dev]: passing of the with_llr switch to the HybridMonteCarlo template class.
- Code tested and running successfully.
--------------

Commit:
-- C++ -------
- [Dev]: First phase of the LLR insertion.
- Modification of the various Makefile as need to have both the test codes and driver codes.
- Creation of an additional class for handling the LLR input and computation.
- Code tested: compiles and runs successfully.
--------------

Commit:
-- C++ -------
- [Dev]: Implementation of the constrained action for the LLR.
- Implementation of the testcodes.
- Creation of an additional class for handling the LLR input and computation.
- Extension of the ActionBase.h for getters and setters to retrieve llrparms structure.
- Log colors activated.
- Various print statements in WilsonGaugeAction.h.
- Readjustment and bug correction in LLRGaugeAction.h in the derive term and implementation
  of three actions computation, S(a), S(beta) and S(no beta). The S(no beta) is the one
  being used the others are for completeness and checking and may be used from Integrator.h.
- Three tests Sp(2n) LLR_HMC_Sp2n_WilsonGauge.cc, Su(n) LLR_HMC_SUn_WilsonGauge.cc and
  HMC_SUn_WilsonGauge.cc. all passed the tests.
- Several building scripts to ease of configuration and building in various cases.
- Extension of the llr_hmc.cc class used for the LLR.
- HMC has been integrated with the LLR logic for the switch with_llr.
- Code tested: compiles and runs successfully the Metropolis seems to work correctly.
--------------

Commit:
- [LLR-dev]:
- [Test]: modification of the testcode Test_llr_contrained_action.cc to include setters/ getters for both the action and the plaquette.
- Added a few comment and reorganise the code structure and order.
- [Log]: implemented a new specific logger for LLR called GridLogLLR to be analogoeus with the otehr logger. This can be called for when LLR is activated.
- [LLR-fev]: migration from GridLogIntegrator to GridLogLLR in LLRGaugeAction.h.
- [LLR-dev]: minor code edit in llr_hmc.cc.
- [LLR-dev]: Correction of the logic in the reversability in HMC.h. The reversbility now works.
- [LLR-test]: Tested various observables using codes in previous papers.
--------------

Commit:
- [LLR-dev]:Added various scripts building, testing, examples, and running LLR and Grid ./scripts/llr_tools
--------------

Commit:
- [LLR-dev]:Cleanup in HMC.h
- Renaming of some functions in llr_hmc.cc.
- Added Module classes in LLR_HMC_Sp2n_WilsonGauge.cc c ode.
- Cleanup in LLRGaugeAction.h and moved from log GridLogINtegrator to GridLogLLR.
- Cleanup on some the scripts.
- Tidyup in Test_llr_contrained_action.cc test code and start of the testing.
--------------

Commit:
- [LLR-dev]:Cleanup in HMC.h
- Removed all of the print statements used for debugging.
- Removed unnecessary variables like H0_llr H1_llr that were used for debugging and verifications and sanity checks. I have kept Grid:Reald H0 = 0.0 because I really want to have those variables initialised that way. Also because there is a if block.
--------------

Commit:
- [LLR-dev]: Integration of the Assert statement for Test_llr_contrained_action.cc.
- Cleanup in HMC.h
- Cleanup in Integrator.h.
- Removed all of print statements used for debugging.
- Removed unnecessary variables like H0_llr H1_llr that were used for debugging and verifications and sanity checks. I have kept Grid:RealD H0 = 0.0 because I really want to have those variables initialised that way. Also because there is a if block.
- Added two additional variables in typedef struct llrparams for the action S and the plaquette plaq.
- Update with the latest for LLR_HMC_Sp2n_WilsonGauge.cc, LLR_HMC_SUn_WilsonGauge.cc codes for llrparams.
- Insertion of the template class for logger in LLR_HMC_SUn_WilsonGauge.cc.
- Finalizing statements have been added in LLR_HMC_Sp2n_WilsonGauge.cc and soon also in LLR_HMC_SUn_WilsonGauge.cc.
--------------

Commit:
- [LLR-dev]: Update of LLR_HMC_SUn_WilsonGauge.cc with the hmc struct usage.
- Update of the launching scripts Run_LLR_HMC_SUn_WilsonGauge_nodes001_mpi01-01-01-04.sh and Run_Test_llr_contrained_action_nodes001_mpi01-01-01-04.sh for command line driven launches.
- Added an additional and needed constructor in llr_hmc.cc.
- In HMC_SUn_WilsonGauge.cc added a logger that also includes Topological charge computation.
- Code editing and added MDsteps in the parameters list for logging. Also minor code editing in the same file.
- Similarly for LLR_HMC_SUn_WilsonGauge.cc.
- Updated the runners scripts.
- Update in Test_llr_contrained_action.cc to include the MDstep in the logger.
--------------

Commit:
- [LLR-dev]: Minor code edit in llr_hmc.h.
--------------

fbonnet08 committed:
Merge branch 'develop' into develop

edbennett commented on Grid/log/Log.cc:
```diff
@@ -85,8 +86,9 @@ void GridLogConfigure(std::vector<std::string> &logstreams) {
   GridLogDslash.Active(0);
   GridLogIntegrator.Active(1);
-  GridLogColours.Active(0);
+  GridLogColours.Active(1);
```
Please revert this change; it isn't related to the functionality we're touching.

Based on line 105 below, the intended mechanism for enabling log colouring seems to be passing `"Colours"` to `GridLogConfigure`.

edbennett commented on Grid/qcd/action/gauge/Gauge.h:
```diff
@@ -42,26 +45,33 @@ typedef WilsonGaugeAction<PeriodicGimplD>          WilsonGaugeActionD;
 typedef WilsonGaugeAction<SpPeriodicGimplF>        SpWilsonGaugeActionF;
 typedef WilsonGaugeAction<SpPeriodicGimplD>        SpWilsonGaugeActionD;
+
+//typedef LLRWilsonGaugeAction<PeriodicGimplR>          LLRWilsonGaugeActionR;
```
Don't introduce new commented-out code. Either revert this change, or uncomment it, and also add `D` and `F`-suffixed versions.

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+//
+// Created by dc-bonn2 on 9/17/25.
+//
```
Try to make this consistent with the header used in other files, e.g. `Log.h`.

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+#include <Grid/qcd/llr_hmc/llr_hmc.h>
+
+#ifndef GRID_TELOS_DEVEL_LLRGAUGEACTION_H
+#define GRID_TELOS_DEVEL_LLRGAUGEACTION_H
```
This symbol doesn't need to be namespaces to `TELOS`, or really to `DEVEL`. I'd suggest `QCD_LLR_GAUGE_ACTION_H` to be consistent with the symbol defined in `WilsonGaugeAction.h`.

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+
+NAMESPACE_BEGIN(Grid);
+
+//template<class Gimpl>
```
Don't introduce commented out code.

Commit:
- [LLR-dev]: Updated the logMessage for some of the run codes. LLR_HMC_SUn_WilsonGauge.cc, LLR_HMC_Sp2n_WilsonGauge.cc, Test_llr_contrained_action.cc and HMC_SUn_WilsonGauge.cc.
--------------

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+    virtual RealD GActionS(const GaugeAction &p_GaugeAction, const GaugeField &U) {
+        RealD gauge_action = 0.0;
+        //gauge_action =  p_GaugeAction.S(U);
+        std::cout << GridLogLLR << B_GREEN << "Gauge action LLRGaugeAction      ----->: "<< gauge_action << C_RESET <<std::endl;
```
This seems like an elaborate way of printing out zero.

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+                beta(beta_),
+                p_llrparams_s(llrparams_){};
+
+    virtual std::string action_name() {return "LLRGaugeAction";}
```
This should vary based on the underlying non-LLR gauge action. E.g. something like `LLR_WilsonGaugeAction`, `LLR_PlaqPlusRectangleAction`.

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+    // instantiating the llr_hmc helper class via the namespace LLR
+    namespace_LLR::llrparams* p_llrparams_s;
+    namespace_LLR::llr_hmc* p_llr_hmc_o = new namespace_LLR::llr_hmc();
+    /////////////////////////////////////////////////////////////
+    /// [Constructors]
+    /////////////////////////////////////////////////////////////
+    explicit LLRGaugeAction(RealD beta_):
+                beta(beta_){};
+
+    explicit LLRGaugeAction(namespace_LLR::llrparams* llrparams_, RealD beta_):
+                beta(beta_),
+                p_llrparams_s(llrparams_){};
```
Based on our previous discussions over the last few weeks, $\beta$ should not factor into the LLR gauge action—appearances of it must be replaced with either $a$ or $1$. Why do we allow it to be set at all?

edbennett commented on HMC/HMC_SUn_WilsonGauge.cc:
```diff
@@ -0,0 +1,383 @@
+/*************************************************************************************
```
This file seems superfluous—can it be removed?

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+    virtual void refresh(const GaugeField &U, GridSerialRNG &sRNG, GridParallelRNG &pRNG){};
+
+    virtual RealD S(const GaugeField &U) {
+        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
+        RealD vol = U.Grid()->gSites();
+        //RealD action = beta * (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
+        RealD action = (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
+        std::cout << GridLogLLR << B_GREEN << "plaq(nb) LLRGaugeAction         ----->: "<< plaq   << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_GREEN << "Vol(nb)  LLRGaugeAction         ----->: "<< vol    << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_GREEN << "S(nb)    LLRGaugeAction         ----->: "<< action << C_RESET <<std::endl;
+        return action;
+    };
```
We already had a `using Action<GaugeField>::S` above, so this declaration seems like it shouldn't be needed.

If it's not possible to expose the `S` from the template parameter, then we should replace the function body with a call to it, since the current form means that even if we template on something other than `WilsonGaugeAction`, we'll still end up with the action actually secretly being a `WilsonGaugeAction`.

edbennett replied:
Obviously `beta` would need to be set to `1`, but that can be done while addressing the previous comment about explicit `beta.

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+        RealD vol = U.Grid()->gSites();
+        //RealD action = beta * (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
+        RealD action = (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
+        std::cout << GridLogLLR << B_GREEN << "plaq(nb) LLRGaugeAction         ----->: "<< plaq   << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_GREEN << "Vol(nb)  LLRGaugeAction         ----->: "<< vol    << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_GREEN << "S(nb)    LLRGaugeAction         ----->: "<< action << C_RESET <<std::endl;
```
Now that we ostensibly have the code working, this level of logging is absurdly verbose; can it be removed?

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+        return action;
+    };
+
+    virtual RealD Sa(const GaugeField &U) {
+        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
+        RealD vol = U.Grid()->gSites();
+        RealD action = p_llrparams_s->a * (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
+        std::cout << GridLogLLR << B_MAGENTA << "plaq(a) LLRGaugeAction          ----->: "<< plaq   << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_MAGENTA << "Vol(a)  LLRGaugeAction          ----->: "<< vol    << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_MAGENTA << "S(a)    LLRGaugeAction          ----->: "<< action << C_RESET <<std::endl;
+        return action;
+    };
```
Can this not be replaced with `return p_llrparams_s->a * S(U);`?

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+        return action;
+    };
+
+    virtual RealD Sbeta(const GaugeField &U) {
+        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
+        RealD vol = U.Grid()->gSites();
+        RealD action = beta * (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
+        std::cout << GridLogLLR << B_CYAN << "plaq(beta) LLRGaugeAction       ----->: "<< plaq   << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_CYAN << "Vol(beta)  LLRGaugeAction       ----->: "<< vol    << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_CYAN << "S(beta)    LLRGaugeAction       ----->: "<< action << C_RESET <<std::endl;
+        return action;
+    };
```
Can this not be replaced with `return beta * S(U);`?

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+
+            PokeIndex<LorentzIndex>(dSdU, dSdU_mu, mu);
+        }
+}
```
Fix indent level

edbennett commented on Grid/qcd/action/gauge/LLRGaugeAction.h:
```diff
@@ -0,0 +1,162 @@
+        RealD factor = 0.5  / RealD(Nc);//0.5 * beta / RealD(Nc); //0.5 *
+        GridBase *grid = U.Grid();
+
+        std::cout << GridLogLLR << B_YELLOW   << "factor(deriv) LLRGaugeAction    ----->: "<< factor << C_RESET <<std::endl;
+
+        GaugeLinkField dSdU_mu(grid);
+        std::vector<GaugeLinkField> Umu(Nd, grid);
+
+        for (int mu = 0; mu < Nd; mu++) {
+            Umu[mu] = PeekIndex<LorentzIndex>(U, mu);
+        }
+
+        // Get the average Plaquette to compute S_0
+        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
+        RealD vol = U.Grid()->gSites();
+        // now construct the S_0
+        RealD S0_plaq_ = 6 * vol * plaq;
+        // hard coded value for now TODO: remove this value when properly setup
+        RealD S0_ = p_llrparams_s->S0; //173088.00000;
+
+        std::cout << GridLogLLR << B_RED     << "Vol(deriv)      LLRGaugeAction  ----->: "<< vol               << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_CYAN    << "plaq(deriv)     LLRGaugeAction  ----->: "<< plaq              << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_YELLOW  << "S0_plaq_(deriv) LLRGaugeAction  ----->: "<< S0_plaq_          << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_MAGENTA << "p->S0(deriv)    LLRGaugeAction  ----->: "<< p_llrparams_s->S0 << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_YELLOW  << "p->a(deriv)     LLRGaugeAction  ----->: "<< p_llrparams_s->a  << C_RESET <<std::endl;
+        std::cout << GridLogLLR << B_YELLOW  << "p->dS(deriv)    LLRGaugeAction  ----->: "<< p_llrparams_s->dS << C_RESET <<std::endl;
+
+        //GaugeLinkField new_dSdU_mu(grid);
+        RealD action_u = 0.0;
+        action_u = S(U);
+        for (int mu = 0; mu < Nd; mu++) {
+            // Getting the action for when action
+            // Staple in direction mu
+            WilsonLoops<Gimpl>::Staple(dSdU_mu, Umu, mu);
+
+            dSdU_mu *= ( p_llrparams_s->a + (action_u - p_llrparams_s->S0) / (p_llrparams_s->dS * p_llrparams_s->dS) );
+            dSdU_mu = Ta(Umu[mu] * dSdU_mu) * factor;
+
+            PokeIndex<LorentzIndex>(dSdU, dSdU_mu, mu);
+        }
```
Similarly to the comment about `S` above, this copy-paste of `WilsonGaugeAction` means that the template parameter is ignored, and this always uses the Wilson action.

Unlike the `S` case, we can't just directly call the `deriv` method of the template action as-is, because we need to do the multiplication by the LLR factor before performing the projection on line 136.

Please can you think about ways around this, that will allow us to correctly template on the gauge action?

edbennett commented on Grid/qcd/action/gauge/WilsonGaugeAction.h:
```diff
@@ -29,6 +29,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 See the full license in the file "LICENSE" in the top level distribution
 directory
 *************************************************************************************/
+#include <Grid/qcd/llr_hmc/llr_hmc.h>
```
I would suggest that all of the changes in this file should be reverted.

edbennett commented on Grid/qcd/action/ActionBase.h:
```diff
@@ -98,8 +99,18 @@ class Action
+    /////////////////////////////
+    // LLR methods
+    /////////////////////////////
+    /// [Sturct]
```
Please run a spell-check over your comments.

edbennett commented on Grid/qcd/action/ActionBase.h:
```diff
@@ -98,8 +99,18 @@ class Action
- 
-  /////////////////////////////////////////////////////////////
+    /////////////////////////////
+    // LLR methods
```
We should make it clear to downstream developers that these new attributes and methods aren't needed unless you're implementing the LLR.

Some options for this:
- Write a more detailed comments at this point describing the use case
- Create a subclass or template specialisation for LLR actions that has these attributes while we leave the superclass untouched.
- Use templating to make the LLR attributes invisible on non-LLR actions.

edbennett commented on Grid/qcd/action/ActionBase.h:
```diff
@@ -98,8 +99,18 @@ class Action
+    virtual void set_struct_llrparams(namespace_LLR::llrparams* llrparams_s_in) {
+        p_llrparams_s = llrparams_s_in;
+    }
+    /////////////////////////////////////////////////////////////
```
Restore the indentation to the original level here.

edbennett commented on Grid/qcd/action/ActionBase.h:
```diff
@@ -28,6 +28,7 @@ See the full license in the file "LICENSE" in the top level distribution
 directory
 *************************************************************************************/
 			   /*  END LEGAL */
+#include <Grid/qcd/llr_hmc/llr_hmc.h>
```
Iff it is essential to include new headers, then this should be done inside the `#ifndef` block.

However, we should think carefully about where the headers are placed, so we don't create the appearance of circular references.

`Grid/qcd/hmc` uses concepts from `Grid/qcd/action`, so having `Grid/qcd/action` refer to the LLR analogue `Grid/qcd/llr_hmc` is concerning.

edbennett commented on Grid/qcd/hmc/integrators/Integrator.h:
```diff
@@ -34,6 +34,8 @@ directory
 
 #include <memory>
 
+#include "Grid/qcd/llr_hmc/llr_hmc.h"
```
Should this not be `#include <>` rather than `#include ""`?

edbennett commented on Grid/qcd/hmc/integrators/Integrator.h:
```diff
@@ -160,7 +168,7 @@ class Integrator {
       Real impulse_max = force_max * ep * HMC_MOMENTUM_DENOMINATOR;    
 
       as[level].actions.at(a)->deriv_log(force_abs,force_max,impulse_abs,impulse_max);
-      
+
```
Avoid making unneeded whitespace changes.

edbennett commented on Grid/qcd/hmc/integrators/Integrator.h:
```diff
@@ -468,8 +476,10 @@ class Integrator {
         Hterm = as[level].actions.at(actionID)->S(Smearer);
-   	        as[level].actions.at(actionID)->S_timer_stop();
+
+            as[level].actions.at(actionID)->S_timer_stop();
```
Avoid making unneeded whitespace changes.

edbennett commented on Grid/qcd/hmc/integrators/Integrator.h:
```diff
@@ -480,6 +490,58 @@ class Integrator {
   }
 
+  // LLR- implementation
+  RealD S_llr(Field& U) {
```
This function should only be callable if the integrator is working on LLR. This could be achieved via templating.

edbennett replied:
This function also has a lot of lines in common with the function `S(Field& U)`. Can the two functions be merged, perhaps as a function template, to avoid this repetition?

edbennett commented on Grid/qcd/hmc/integrators/Integrator.h:
```diff
@@ -495,7 +557,56 @@ class Integrator {
 
-  RealD Sinitial(Field& U) 
+    // LLR- implementation
+  RealD Sinitial_llr(Field& U) {
```
As above for `S`: can this be merged with `Sinitial(Field& U)`?

edbennett commented on Grid/qcd/hmc/integrators/Integrator.h:
```diff
@@ -552,6 +663,7 @@ class Integrator {
 
   }
 
+private:
```
Remove this

edbennett commented on Grid/qcd/hmc/GenericHMCrunner.h:
```diff
@@ -78,6 +81,15 @@ class HMCWrapperTemplate: public HMCRunnerBase<ReaderClass> {
   void ReadCommandLine(int argc, char **argv) {
     std::string arg;
 
+    if (GridCmdOptionExists(argv, argv + argc, "--enable-llr")) {
+        arg = GridCmdOptionPayload(argv, argv + argc, "--enable-llr");
+
+        with_llr = true;
+
+        std::cout <<GridLogMessage << " with_llr         ---------->: "<<with_llr<<std::endl;
+        std::cout <<GridLogMessage << " GenericHMCrunner ---------->: --enable-llr"<<std::endl;
+    }
+
```
We will never want the same application to be usable for both LLR and non-LLR—they are too divergent. As such we don't want a run-time switch for this.

My original thinking was that we would have a _configure_-time switch to decide whether the LLR templates are instantiated. I'm not sure how necessary that is with the current design.

edbennett commented on Grid/qcd/hmc/GenericHMCrunner.h:
```diff
@@ -47,6 +47,9 @@ template <class Implementation,
           class RepresentationsPolicy = NoHirep, class ReaderClass = XmlReader>
 class HMCWrapperTemplate: public HMCRunnerBase<ReaderClass> {
 public:
+  // The with LLR boolean switch
```
This comment gives no information to the reader beyond what is in the code in the following line. Either expand it to explain the intention and effects of `with_llr`, or remove the comment entirely.

edbennett commented on Grid/qcd/hmc/GenericHMCrunner.h:
```diff
@@ -47,6 +47,9 @@ template <class Implementation,
 class HMCWrapperTemplate: public HMCRunnerBase<ReaderClass> {
 public:
+  // The with LLR boolean switch
+  bool with_llr = false;
```
I don't believe that there are are any situations where we want to take an HMC from being LLR to being non-LLR or vice-versa, so this should probably be `const`.

edbennett commented on Grid/qcd/hmc/GenericHMCrunner.h:
```diff
@@ -90,7 +102,7 @@ class HMCWrapperTemplate: public HMCRunnerBase<ReaderClass> {
         exit(1);
       }
       Parameters.StartingType = arg;
-      std::cout <<GridLogMessage << " GenericHMCrunner --StartingType "<<arg<<std::endl;
+        std::cout <<GridLogMessage << " GenericHMCrunner --StartingType "<<arg<<std::endl;
```
Avoid changing spacing in code that we are not otherwise modifying.

edbennett commented on Grid/qcd/hmc/GenericHMCrunner.h:
```diff
@@ -195,10 +207,17 @@ class HMCWrapperTemplate: public HMCRunnerBase<ReaderClass> {
+
+      std::cout <<GridLogMessage << "with_llr b4       ---------->: "<< with_llr     <<std::endl;
+      std::cout <<GridLogMessage << "HybridMonteCarlo call ------>: "<< __FUNCTION__ <<std::endl;
+
```
If this is kept, it should probably be as `GridLogDebug`, but more likely it should be removed.

edbennett commented on Grid/qcd/hmc/GenericHMCrunner.h:
```diff
@@ -223,14 +242,20 @@ template <template <typename, typename, typename> class Integrator>
 				     HMCWrapperTemplate<PeriodicGimplR, Integrator, RepresentationsPolicy>;
 
 // sp2n
-
```
Avoid changing spacing in code that we are not otherwise editing

edbennett commented on Grid/qcd/hmc/GenericHMCrunner.h:
```diff
@@ -240,7 +265,6 @@ using GenericSpHMCRunnerHirep =
                      HMCWrapperTemplate<SpPeriodicGimplR, Integrator, RepresentationsPolicy>;
 
 
-
```
Avoid changing spacing in code that we are not otherwise editing

edbennett reviewed (changes requested):
Hi Frederic

I've been on trains this afternoon, and it was hard to get a seat, so I haven't been able to review 100% of the code yet. But in the mean time, please see below my suggestions on the code I've read thus far.

In addition, there are many places where you've introduced new one- or two-letter variables. While this does match the style in which Grid is written, which is good, we have found these to be very difficult to read unless you are reading exactly the same references that the original developers were. For each such variable you introduce, please can you add a 1–2-line comment explaining the meaning of the variable, and where it is defined? Similarly for functionality, it will be immensely valuable to those reading the code in future if each block of functionality (function, or group of lines added to an existing function) has a comment explaining the reasoning behind it (or even just an arXiv number and equation number where someone else has written the explanation).

I'm at an event all day tomorrow so won't have much time to progress a review; it's possible I'll be able to have another look on the train on the way home.

fbonnet08 commented:
Hi Ed,

Thanks for the update, I will make the modifications tomorrow morning on a
fresh start and carefully.

I tried to clean up as much as possible.

The one in Gauge.h certainly needs to be removed.

I like to put things like

+/*************************************************************************************

to air out the code and ease the readability it really  helps me, but
I can remove it.

I will also restore all of the indenting with the Grid and match it.

I make the modifications and push onto the repo. Also the
Accelerator.h (related to HIP) a bug that needs to be corrected

must be in there. I was really confused today as I thought it was
reverted. This is important because if it is not correct

it will fail on AMD GPUs. This is what we saw at the AMD hackathon.

Frederic.
--
Frederic Bonnet
***@***.***

=========================================================
This  email  message  is  intended   only for the addressee(s)
and contains  information  which may be confidential  and / or
copyright. If you are not the intended recipient please do not
read, save, forward, disclose, or copy the contents of this email.
If this email has  been sent to  you  in error, please delete this
email  and  any  copies  or  links  to  this  email  completely  and
immediately from your system.  No representation  is  made
that  this email is free of viruses. Virus scanning  is recommended
and  is the responsibility of the recipient.
=========================================================



On Wed, Nov 26, 2025, 21:45 Ed Bennett ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Hi Frederic
>
> I've been on trains this afternoon, and it was hard to get a seat, so I
> haven't been able to review 100% of the code yet. But in the mean time,
> please see below my suggestions on the code I've read thus far.
>
> In addition, there are many places where you've introduced new one- or
> two-letter variables. While this does match the style in which Grid is
> written, which is good, we have found these to be very difficult to read
> unless you are reading exactly the same references that the original
> developers were. For each such variable you introduce, please can you add a
> 1–2-line comment explaining the meaning of the variable, and where it is
> defined? Similarly for functionality, it will be immensely valuable to
> those reading the code in future if each block of functionality (function,
> or group of lines added to an existing function) has a comment explaining
> the reasoning behind it (or even just an arXiv number and equation number
> where someone else has written the explanation).
>
> I'm at an event all day tomorrow so won't have much time to progress a
> review; it's possible I'll be able to have another look on the train on the
> way home.
> ------------------------------
>
> In Grid/log/Log.cc
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2565506633>
> :
>
> > @@ -85,8 +86,9 @@ void GridLogConfigure(std::vector<std::string> &logstreams) {
>    GridLogPerformance.Active(0);
>    GridLogDslash.Active(0);
>    GridLogIntegrator.Active(1);
> -  GridLogColours.Active(0);
> +  GridLogColours.Active(1);
>
> Please revert this change; it isn't related to the functionality we're
> touching.
>
> Based on line 105 below, the intended mechanism for enabling log colouring
> seems to be passing "Colours" to GridLogConfigure.
> ------------------------------
>
> In Grid/qcd/action/gauge/Gauge.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2565513216>
> :
>
> > @@ -42,26 +45,33 @@ typedef WilsonGaugeAction<PeriodicGimplD>          WilsonGaugeActionD;
>  typedef WilsonGaugeAction<SpPeriodicGimplR>        SpWilsonGaugeActionR;
>  typedef WilsonGaugeAction<SpPeriodicGimplF>        SpWilsonGaugeActionF;
>  typedef WilsonGaugeAction<SpPeriodicGimplD>        SpWilsonGaugeActionD;
> +
> +//typedef LLRWilsonGaugeAction<PeriodicGimplR>          LLRWilsonGaugeActionR;
>
> Don't introduce new commented-out code. Either revert this change, or
> uncomment it, and also add D and F-suffixed versions.
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2565518361>
> :
>
> > +//
> +// Created by dc-bonn2 on 9/17/25.
> +//
>
> Try to make this consistent with the header used in other files, e.g.
> Log.h.
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2565536682>
> :
>
> > +#ifndef GRID_TELOS_DEVEL_LLRGAUGEACTION_H
> +#define GRID_TELOS_DEVEL_LLRGAUGEACTION_H
>
> This symbol doesn't need to be namespaces to TELOS, or really to DEVEL.
> I'd suggest QCD_LLR_GAUGE_ACTION_H to be consistent with the symbol
> defined in WilsonGaugeAction.h.
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2565539170>
> :
>
> > @@ -0,0 +1,162 @@
> +//
> +// Created by dc-bonn2 on 9/17/25.
> +//
> +
> +#include <Grid/qcd/llr_hmc/llr_hmc.h>
> +
> +#ifndef GRID_TELOS_DEVEL_LLRGAUGEACTION_H
> +#define GRID_TELOS_DEVEL_LLRGAUGEACTION_H
> +
> +NAMESPACE_BEGIN(Grid);
> +
> +//template<class Gimpl>
>
> Don't introduce commented out code.
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566430795>
> :
>
> > +        RealD gauge_action = 0.0;
> +        //gauge_action =  p_GaugeAction.S(U);
> +        std::cout << GridLogLLR << B_GREEN << "Gauge action LLRGaugeAction      ----->: "<< gauge_action << C_RESET <<std::endl;
>
> This seems like an elaborate way of printing out zero.
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566434043>
> :
>
> > +    using Action<GaugeField>::get_struct_llrparams;
> +
> +    // instantiating the llr_hmc helper class via the namespace LLR
> +    namespace_LLR::llrparams* p_llrparams_s;
> +    namespace_LLR::llr_hmc* p_llr_hmc_o = new namespace_LLR::llr_hmc();
> +    /////////////////////////////////////////////////////////////
> +    /// [Constructors]
> +    /////////////////////////////////////////////////////////////
> +    explicit LLRGaugeAction(RealD beta_):
> +                beta(beta_){};
> +
> +    explicit LLRGaugeAction(namespace_LLR::llrparams* llrparams_, RealD beta_):
> +                beta(beta_),
> +                p_llrparams_s(llrparams_){};
> +
> +    virtual std::string action_name() {return "LLRGaugeAction";}
>
> This should vary based on the underlying non-LLR gauge action. E.g.
> something like LLR_WilsonGaugeAction, LLR_PlaqPlusRectangleAction.
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566436349>
> :
>
> > +    explicit LLRGaugeAction(RealD beta_):
> +                beta(beta_){};
> +
> +    explicit LLRGaugeAction(namespace_LLR::llrparams* llrparams_, RealD beta_):
> +                beta(beta_),
> +                p_llrparams_s(llrparams_){};
>
> Based on our previous discussions over the last few weeks, $\beta$ should
> not factor into the LLR gauge action—appearances of it must be replaced
> with either $a$ or $1$. Why do we allow it to be set at all?
> ------------------------------
>
> In HMC/HMC_SUn_WilsonGauge.cc
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566448205>
> :
>
> > @@ -0,0 +1,383 @@
> +/*************************************************************************************
>
> This file seems superfluous—can it be removed?
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566459543>
> :
>
> > +    virtual RealD S(const GaugeField &U) {
> +        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
> +        RealD vol = U.Grid()->gSites();
> +        //RealD action = beta * (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
> +        RealD action = (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
> +        std::cout << GridLogLLR << B_GREEN << "plaq(nb) LLRGaugeAction         ----->: "<< plaq   << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_GREEN << "Vol(nb)  LLRGaugeAction         ----->: "<< vol    << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_GREEN << "S(nb)    LLRGaugeAction         ----->: "<< action << C_RESET <<std::endl;
> +        return action;
> +    };
>
> We already had a using Action<GaugeField>::S above, so this declaration
> seems like it shouldn't be needed.
>
> If it's not possible to expose the S from the template parameter, then we
> should replace the function body with a call to it, since the current form
> means that even if we template on something other than WilsonGaugeAction,
> we'll still end up with the action actually secretly being a
> WilsonGaugeAction.
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566461269>
> :
>
> > +        std::cout << GridLogLLR << B_GREEN << "plaq(nb) LLRGaugeAction         ----->: "<< plaq   << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_GREEN << "Vol(nb)  LLRGaugeAction         ----->: "<< vol    << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_GREEN << "S(nb)    LLRGaugeAction         ----->: "<< action << C_RESET <<std::endl;
>
> Now that we ostensibly have the code working, this level of logging is
> absurdly verbose; can it be removed?
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566472873>
> :
>
> > +    virtual RealD Sa(const GaugeField &U) {
> +        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
> +        RealD vol = U.Grid()->gSites();
> +        RealD action = p_llrparams_s->a * (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
> +        std::cout << GridLogLLR << B_MAGENTA << "plaq(a) LLRGaugeAction          ----->: "<< plaq   << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_MAGENTA << "Vol(a)  LLRGaugeAction          ----->: "<< vol    << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_MAGENTA << "S(a)    LLRGaugeAction          ----->: "<< action << C_RESET <<std::endl;
> +        return action;
> +    };
>
> Can this not be replaced with return p_llrparams_s->a * S(U);?
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566474395>
> :
>
> > +    virtual RealD Sbeta(const GaugeField &U) {
> +        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
> +        RealD vol = U.Grid()->gSites();
> +        RealD action = beta * (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
> +        std::cout << GridLogLLR << B_CYAN << "plaq(beta) LLRGaugeAction       ----->: "<< plaq   << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_CYAN << "Vol(beta)  LLRGaugeAction       ----->: "<< vol    << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_CYAN << "S(beta)    LLRGaugeAction       ----->: "<< action << C_RESET <<std::endl;
> +        return action;
> +    };
>
> Can this not be replaced with return beta * S(U);?
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566479810>
> :
>
> > +        std::cout << GridLogLLR << B_YELLOW  << "p->dS(deriv)    LLRGaugeAction  ----->: "<< p_llrparams_s->dS << C_RESET <<std::endl;
> +
> +        //GaugeLinkField new_dSdU_mu(grid);
> +        RealD action_u = 0.0;
> +        action_u = S(U);
> +        for (int mu = 0; mu < Nd; mu++) {
> +            // Getting the action for when action
> +            // Staple in direction mu
> +            WilsonLoops<Gimpl>::Staple(dSdU_mu, Umu, mu);
> +
> +            dSdU_mu *= ( p_llrparams_s->a + (action_u - p_llrparams_s->S0) / (p_llrparams_s->dS * p_llrparams_s->dS) );
> +            dSdU_mu = Ta(Umu[mu] * dSdU_mu) * factor;
> +
> +            PokeIndex<LorentzIndex>(dSdU, dSdU_mu, mu);
> +        }
> +}
>
> Fix indent level
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566498563>
> :
>
> > +    virtual void deriv(const GaugeField &U, GaugeField &dSdU) {
> +        // not optimal implementation FIXME
> +        // extend Ta to include Lorentz indexes
> +        std::cout << GridLogLLR << B_GREEN   << "(deriv)     LLRGaugeAction"              << C_RESET   << std::endl;
> +        std::cout << GridLogLLR << B_MAGENTA << "beta(deriv) LLRGaugeAction      ----->: "<< beta      << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_MAGENTA << "Nc(deriv)   LLRGaugeAction      ----->: "<< RealD(Nc) << C_RESET <<std::endl;
> +
> +        RealD factor = 0.5  / RealD(Nc);//0.5 * beta / RealD(Nc); //0.5 *
> +        GridBase *grid = U.Grid();
> +
> +        std::cout << GridLogLLR << B_YELLOW   << "factor(deriv) LLRGaugeAction    ----->: "<< factor << C_RESET <<std::endl;
> +
> +        GaugeLinkField dSdU_mu(grid);
> +        std::vector<GaugeLinkField> Umu(Nd, grid);
> +
> +        for (int mu = 0; mu < Nd; mu++) {
> +            Umu[mu] = PeekIndex<LorentzIndex>(U, mu);
> +        }
> +
> +        // Get the average Plaquette to compute S_0
> +        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
> +        RealD vol = U.Grid()->gSites();
> +        // now construct the S_0
> +        RealD S0_plaq_ = 6 * vol * plaq;
> +        // hard coded value for now TODO: remove this value when properly setup
> +        RealD S0_ = p_llrparams_s->S0; //173088.00000;
> +
> +        std::cout << GridLogLLR << B_RED     << "Vol(deriv)      LLRGaugeAction  ----->: "<< vol               << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_CYAN    << "plaq(deriv)     LLRGaugeAction  ----->: "<< plaq              << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_YELLOW  << "S0_plaq_(deriv) LLRGaugeAction  ----->: "<< S0_plaq_          << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_MAGENTA << "p->S0(deriv)    LLRGaugeAction  ----->: "<< p_llrparams_s->S0 << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_YELLOW  << "p->a(deriv)     LLRGaugeAction  ----->: "<< p_llrparams_s->a  << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_YELLOW  << "p->dS(deriv)    LLRGaugeAction  ----->: "<< p_llrparams_s->dS << C_RESET <<std::endl;
> +
> +        //GaugeLinkField new_dSdU_mu(grid);
> +        RealD action_u = 0.0;
> +        action_u = S(U);
> +        for (int mu = 0; mu < Nd; mu++) {
> +            // Getting the action for when action
> +            // Staple in direction mu
> +            WilsonLoops<Gimpl>::Staple(dSdU_mu, Umu, mu);
> +
> +            dSdU_mu *= ( p_llrparams_s->a + (action_u - p_llrparams_s->S0) / (p_llrparams_s->dS * p_llrparams_s->dS) );
> +            dSdU_mu = Ta(Umu[mu] * dSdU_mu) * factor;
> +
> +            PokeIndex<LorentzIndex>(dSdU, dSdU_mu, mu);
> +        }
>
> Similarly to the comment about S above, this copy-paste of
> WilsonGaugeAction means that the template parameter is ignored, and this
> always uses the Wilson action.
>
> Unlike the S case, we can't just directly call the deriv method of the
> template action as-is, because we need to do the multiplication by the LLR
> factor before performing the projection on line 136.
>
> Please can you think about ways around this, that will allow us to
> correctly template on the gauge action?
> ------------------------------
>
> In Grid/qcd/action/gauge/WilsonGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566500941>
> :
>
> > @@ -29,6 +29,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
>  See the full license in the file "LICENSE" in the top level distribution
>  directory
>  *************************************************************************************/
> +#include <Grid/qcd/llr_hmc/llr_hmc.h>
>
> I would suggest that all of the changes in this file should be reverted.
> ------------------------------
>
> In Grid/qcd/action/ActionBase.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566503171>
> :
>
> > @@ -98,8 +99,18 @@ class Action
>    virtual RealD S(const GaugeField& U) = 0;                             // evaluate the action
>    virtual RealD Sinitial(const GaugeField& U) { return this->S(U); } ;  // if the refresh computes the action, can cache it. Alternately refreshAndAction() ?
>    virtual void deriv(const GaugeField& U, GaugeField& dSdU) = 0;        // evaluate the action derivative
> -
> -  /////////////////////////////////////////////////////////////
> +    /////////////////////////////
> +    // LLR methods
> +    /////////////////////////////
> +    /// [Sturct]
>
> Please run a spell-check over your comments.
> ------------------------------
>
> In Grid/qcd/action/ActionBase.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566506799>
> :
>
> > @@ -98,8 +99,18 @@ class Action
>    virtual RealD S(const GaugeField& U) = 0;                             // evaluate the action
>    virtual RealD Sinitial(const GaugeField& U) { return this->S(U); } ;  // if the refresh computes the action, can cache it. Alternately refreshAndAction() ?
>    virtual void deriv(const GaugeField& U, GaugeField& dSdU) = 0;        // evaluate the action derivative
> -
> -  /////////////////////////////////////////////////////////////
> +    /////////////////////////////
> +    // LLR methods
>
> We should make it clear to downstream developers that these new attributes
> and methods aren't needed unless you're implementing the LLR.
>
> Some options for this:
>
>    - Write a more detailed comments at this point describing the use case
>    - Create a subclass or template specialisation for LLR actions that
>    has these attributes while we leave the superclass untouched.
>    - Use templating to make the LLR attributes invisible on non-LLR
>    actions.
>
> ------------------------------
>
> In Grid/qcd/action/ActionBase.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566507258>
> :
>
> > @@ -98,8 +99,18 @@ class Action
>    virtual RealD S(const GaugeField& U) = 0;                             // evaluate the action
>    virtual RealD Sinitial(const GaugeField& U) { return this->S(U); } ;  // if the refresh computes the action, can cache it. Alternately refreshAndAction() ?
>    virtual void deriv(const GaugeField& U, GaugeField& dSdU) = 0;        // evaluate the action derivative
> -
> -  /////////////////////////////////////////////////////////////
> +    /////////////////////////////
> +    // LLR methods
> +    /////////////////////////////
> +    /// [Sturct]
> +    namespace_LLR::llrparams* p_llrparams_s;
> +    /// [Getters]
> +    virtual namespace_LLR::llrparams* get_struct_llrparams() {return p_llrparams_s;}
> +    /// [Setters]
> +    virtual void set_struct_llrparams(namespace_LLR::llrparams* llrparams_s_in) {
> +        p_llrparams_s = llrparams_s_in;
> +    }
> +    /////////////////////////////////////////////////////////////
>
> Restore the indentation to the original level here.
> ------------------------------
>
> In Grid/qcd/action/ActionBase.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566512289>
> :
>
> > @@ -28,6 +28,7 @@ See the full license in the file "LICENSE" in the top level distribution
>  directory
>  *************************************************************************************/
>  			   /*  END LEGAL */
> +#include <Grid/qcd/llr_hmc/llr_hmc.h>
>
> Iff it is essential to include new headers, then this should be done
> inside the #ifndef block.
>
> However, we should think carefully about where the headers are placed, so
> we don't create the appearance of circular references.
>
> Grid/qcd/hmc uses concepts from Grid/qcd/action, so having Grid/qcd/action
> refer to the LLR analogue Grid/qcd/llr_hmc is concerning.
> ------------------------------
>
> In Grid/qcd/hmc/integrators/Integrator.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566513563>
> :
>
> > @@ -34,6 +34,8 @@ directory
>
>  #include <memory>
>
> +#include "Grid/qcd/llr_hmc/llr_hmc.h"
>
> Should this not be #include <> rather than #include ""?
> ------------------------------
>
> In Grid/qcd/hmc/integrators/Integrator.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566515261>
> :
>
> > -
> +
>
> Avoid making unneeded whitespace changes.
> ------------------------------
>
> In Grid/qcd/hmc/integrators/Integrator.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566515704>
> :
>
> > +
>          Hterm = as[level].actions.at(actionID)->S(Smearer);
> -   	        as[level].actions.at(actionID)->S_timer_stop();
> +
> +            as[level].actions.at(actionID)->S_timer_stop();
>
> Avoid making unneeded whitespace changes.
> ------------------------------
>
> In Grid/qcd/hmc/integrators/Integrator.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566533132>
> :
>
> > @@ -480,6 +490,58 @@ class Integrator {
>      return H;
>    }
>
> +  // LLR- implementation
> +  RealD S_llr(Field& U) {
>
> This function should only be callable if the integrator is working on LLR.
> This could be achieved via templating.
> ------------------------------
>
> In Grid/qcd/hmc/integrators/Integrator.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566542332>
> :
>
> > @@ -495,7 +557,56 @@ class Integrator {
>      }
>    } Sinitial_hireps{};
>
> -  RealD Sinitial(Field& U)
> +    // LLR- implementation
> +  RealD Sinitial_llr(Field& U) {
>
> As above for S: can this be merged with Sinitial(Field& U)?
> ------------------------------
>
> In Grid/qcd/hmc/integrators/Integrator.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566542788>
> :
>
> > @@ -552,6 +663,7 @@ class Integrator {
>
>    }
>
> +private:
>
> Remove this
> ------------------------------
>
> In Grid/qcd/hmc/GenericHMCrunner.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566546139>
> :
>
> > +    if (GridCmdOptionExists(argv, argv + argc, "--enable-llr")) {
> +        arg = GridCmdOptionPayload(argv, argv + argc, "--enable-llr");
> +
> +        with_llr = true;
> +
> +        std::cout <<GridLogMessage << " with_llr         ---------->: "<<with_llr<<std::endl;
> +        std::cout <<GridLogMessage << " GenericHMCrunner ---------->: --enable-llr"<<std::endl;
> +    }
> +
>
> We will never want the same application to be usable for both LLR and
> non-LLR—they are too divergent. As such we don't want a run-time switch for
> this.
>
> My original thinking was that we would have a *configure*-time switch to
> decide whether the LLR templates are instantiated. I'm not sure how
> necessary that is with the current design.
> ------------------------------
>
> In Grid/qcd/hmc/GenericHMCrunner.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566548055>
> :
>
> > @@ -47,6 +47,9 @@ template <class Implementation,
>            class RepresentationsPolicy = NoHirep, class ReaderClass = XmlReader>
>  class HMCWrapperTemplate: public HMCRunnerBase<ReaderClass> {
>  public:
> +  // The with LLR boolean switch
>
> This comment gives no information to the reader beyond what is in the code
> in the following line. Either expand it to explain the intention and
> effects of with_llr, or remove the comment entirely.
> ------------------------------
>
> In Grid/qcd/hmc/GenericHMCrunner.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566549613>
> :
>
> > @@ -47,6 +47,9 @@ template <class Implementation,
>            class RepresentationsPolicy = NoHirep, class ReaderClass = XmlReader>
>  class HMCWrapperTemplate: public HMCRunnerBase<ReaderClass> {
>  public:
> +  // The with LLR boolean switch
> +  bool with_llr = false;
>
> I don't believe that there are are any situations where we want to take an
> HMC from being LLR to being non-LLR or vice-versa, so this should probably
> be const.
> ------------------------------
>
> In Grid/qcd/hmc/GenericHMCrunner.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566551494>
> :
>
> > -      std::cout <<GridLogMessage << " GenericHMCrunner --StartingType "<<arg<<std::endl;
> +        std::cout <<GridLogMessage << " GenericHMCrunner --StartingType "<<arg<<std::endl;
>
> Avoid changing spacing in code that we are not otherwise modifying.
> ------------------------------
>
> In Grid/qcd/hmc/GenericHMCrunner.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566554070>
> :
>
> > +      std::cout <<GridLogMessage << "sizeof(U)         ---------->: "<< sizeof(U) <<std::endl;
> +      //std::cout <<GridLogMessage << "U                 ---------->: "<< U <<std::endl;
> +
> +      std::cout <<GridLogMessage << "with_llr b4       ---------->: "<< with_llr     <<std::endl;
> +      std::cout <<GridLogMessage << "HybridMonteCarlo call ------>: "<< __FUNCTION__ <<std::endl;
> +
>
> If this is kept, it should probably be as GridLogDebug, but more likely
> it should be removed.
> ------------------------------
>
> In Grid/qcd/hmc/GenericHMCrunner.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566554419>
> :
>
> >  template <class RepresentationsPolicy,
>            template <typename, typename, typename> class Integrator>
>  using GenericHMCRunnerHirep =
>  				     HMCWrapperTemplate<PeriodicGimplR, Integrator, RepresentationsPolicy>;
>
>  // sp2n
> -
>
> Avoid changing spacing in code that we are not otherwise editing
> ------------------------------
>
> In Grid/qcd/hmc/GenericHMCrunner.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566554964>
> :
>
> > @@ -240,7 +265,6 @@ using GenericSpHMCRunnerHirep =
>                       HMCWrapperTemplate<SpPeriodicGimplR, Integrator, RepresentationsPolicy>;
>
>
> -
>
> Avoid changing spacing in code that we are not otherwise editing
> ------------------------------
>
> In Grid/qcd/action/gauge/LLRGaugeAction.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566468964>
> :
>
> > +    virtual RealD S(const GaugeField &U) {
> +        RealD plaq = WilsonLoops<Gimpl>::avgPlaquette(U);
> +        RealD vol = U.Grid()->gSites();
> +        //RealD action = beta * (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
> +        RealD action = (1.0 - plaq) * (Nd * (Nd - 1.0)) * vol * 0.5;
> +        std::cout << GridLogLLR << B_GREEN << "plaq(nb) LLRGaugeAction         ----->: "<< plaq   << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_GREEN << "Vol(nb)  LLRGaugeAction         ----->: "<< vol    << C_RESET <<std::endl;
> +        std::cout << GridLogLLR << B_GREEN << "S(nb)    LLRGaugeAction         ----->: "<< action << C_RESET <<std::endl;
> +        return action;
> +    };
>
> Obviously beta would need to be set to 1, but that can be done while
> addressing the previous comment about explicit `beta.
> ------------------------------
>
> In Grid/qcd/hmc/integrators/Integrator.h
> <https://github.com/telos-collaboration/Grid/pull/14#discussion_r2566541616>
> :
>
> > @@ -480,6 +490,58 @@ class Integrator {
>      return H;
>    }
>
> +  // LLR- implementation
> +  RealD S_llr(Field& U) {
>
> This function also has a lot of lines in common with the function S(Field&
> U). Can the two functions be merged, perhaps as a function template, to
> avoid this repetition?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/telos-collaboration/Grid/pull/14#pullrequestreview-3511581169>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFMCZGWMIJHTC4LSEQPUYD36YNNZAVCNFSM6AAAAACNHZW466VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTKMJRGU4DCMJWHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Commit:
- [LLR-CodeReview]: Code edits for code review.
- ActionBase.h fixed the indenting to match main branch
- Gauge.h removal of the unnecessary comment.
- GenericHMCrunner.h made it all of the previous as it was before.
- Integrator.h added comments, fixed the include statement. Fixed the indenting. Added comments for the action and removed the MemoryManager print statements. Removed the private statement as it is unused.
--------------

Commit:
- [LLR-CodeReview]: Code edits for code review.
- HMC_SUn_WilsonGauge.cc removal of commented code.
- LLR_HMC_Sp2n_WilsonGauge.cc and LLR_HMC_SUn_WilsonGauge.cc remove of commented code as in HMC_SUn_WilsonGauge.cc.
- LLRGaugeAction.h minor code edit.
- Test_llr_contrained_action.cc minor code edit and same code removal as LLR_HMC_Sp2n_WilsonGauge.cc and LLR_HMC_SUn_WilsonGauge.cc.
- WilsonGaugeAction.h removal of all standard probing code and restored to original version from the main branch.
--------------

Commit:
- [LLR-CodeReview]: Code edits for code review.

- HMC.h code edit and line up of indenting.

- Integrator.h minor code edit for std::cout.

- LLRGaugeAction.h removal of the not needed llr_hmc() object instantiation for now line 51. Will be needed at a later stage.

- Test_llr_contrained_action.cc insertion of the possibility to have multiple a values in a loop the variable to change is at
  line 283 at the time of this commit.

  To have
  $> make check
  to work one needs to comment out the line 342
  '''
  s_hmc_params_llr_in = p_llr_hmc_main_o->ReadCommandLine(argc, argv,s_hmc_params_llr_in);
  '''
  The initialized  s_llrparams_in at line 255 will then be used instead of the batch script input
  Run_Test_llr_contrained_action_nodes001_mpi01-01-01-04.sh in test ./scripts/llr_tools/
--------------

edbennett commented:
Many thanks Frederic for your work on this.

For the time being I've merged this into a feature branch `llr-wip`; we'll resume development from here once we have capacity.

fbonnet08 commented:
Hi Ed,

can I still access/pull/push in my own git devel branch on my GitHub
account?

Frederic.

On Mon, Dec 1, 2025 at 2:09 PM Ed Bennett ***@***.***> wrote:

> *edbennett* left a comment (telos-collaboration/Grid#14)
> <https://github.com/telos-collaboration/Grid/pull/14#issuecomment-3596743258>
>
> Many thanks Frederic for your work on this.
>
> For the time being I've merged this into a feature branch llr-wip; we'll
> resume development from here once we have capacity.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/telos-collaboration/Grid/pull/14#issuecomment-3596743258>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFMCZCNWURFRJ2EDCT7B2337RDZRAVCNFSM6AAAAACNHZW466VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKOJWG42DGMRVHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Frederic Bonnet
***@***.***

=========================================================
This  email  message  is  intended   only for the addressee(s)
and contains  information  which may be confidential  and / or
copyright. If you are not the intended recipient please do not
read, save, forward, disclose, or copy the contents of this email.
If this email has  been sent to  you  in error, please delete this
email  and  any  copies  or  links  to  this  email  completely  and
immediately from your system.  No representation  is  made
that  this email is free of viruses. Virus scanning  is recommended
and  is the responsibility of the recipient.
=========================================================

edbennett commented:
Your GitHub account is your own, so yes, you can still make changes in your own forks.

However, I'd recommend you down tools on this project at this point and focus on securing your next position.

Best of luck with your search!
